### PR TITLE
ENT-4595 harmonize core and serialization

### PR DESCRIPTION
**Pull request by blsemo**
This PR brings core, serialization and their deterministic counterparts to a level between OS and ENT.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/internal/ConcurrentBox.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+/**
+ * A [ConcurrentBox] allows the implementation of track() with reduced contention. [concurrent] may be run from several
+ *   threads (which means it MUST be threadsafe!), while [exclusive] stops the world until the tracking has been set up.
+ *   Internally [ConcurrentBox] is implemented simply as a read-write lock.
+ */
+class ConcurrentBox<out T>(val content: T) {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/ConcurrentBox.kt:12]: It seems like nothing references this class here.  Does that mean it should be moved to `node` or something Enterprise specific?

File: core/src/main/kotlin/net/corda/core/internal/utilities/JvmStatsHelper.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package net.corda.core.internal.utilities
+
+import java.lang.management.ManagementFactory
+import java.lang.management.MemoryUsage
+
+/**
+ * Helper method that will prepare JVM stats as a string.
+ */
+class JvmStatsHelper(dumpGcStats: Boolean = true, dumpMemoryStats: Boolean = true, dumpOsStats: Boolean = true) {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/utilities/JvmStatsHelper.kt:9]: Maybe this and the matching test should also be enterprise only?  Maybe we want an explicit `enterprise-common` module that stays in enterprise for classes like this?

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt [hunk 3]
```diff
@@ -436,23 +450,24 @@ object Crypto {
...
-        // TODO Although we handle the above issue here, consider updating to BC 1.61+ which provides a fix.
-        if (signatureScheme == EDDSA_ED25519_SHA512
-                || signatureScheme == SPHINCS256_SHA256
-                || signatureScheme == RSA_SHA256) {
-            signature.initSign(privateKey)
-        } else {
-            // The rest of the algorithms will require a SecureRandom input (i.e., ECDSA or any new algorithm for which
-            // we don't know if it's deterministic).
-            signature.initSign(privateKey, newSecureRandom())
+        return withSignature(signatureScheme) {signature ->
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt:453]: Needs a reformat.

**Review by rick-r3 — COMMENTED** [review]
I'm generally okay with what we "give away" here.  They're actually things that if they are buggy, I'd rather have double the chance to detect that, more than the advantage they give Enterprise.  The advantages are minor and something that constitutes little wins on the extremes of parallelism.

I put a few comments where I think we have `core` classes that have no utility in open source.  Is now the time to create a module to house these in enterprise, since `core` will no longer exist there for common classes (across other modules, like `node`)?

They do not house significant IP though, so if you want to do that _after_ this PR, and delete them from open `core` then, I'm fine with that too.

File: core/src/main/kotlin/net/corda/core/crypto/internal/Instances.kt [hunk 4]
```diff
@@ -1,12 +1,74 @@
...
+    fun <A> withSignature(signatureScheme: SignatureScheme, func: (signature: Signature) -> A): A {
+        val signature = getSignatureInstance(signatureScheme.signatureName, providerMap[signatureScheme.providerName])
+        try {
+            return func(signature)
+        } finally {
+            releaseSignatureInstance(signature)
+        }
+    }
+
+    // Modified for performance in Enterprise.
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/crypto/internal/Instances.kt:25]: remove comment about enterprise

File: core/src/main/kotlin/net/corda/core/internal/utilities/TimingUtils.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package net.corda.core.internal.utilities
+
+import java.util.concurrent.TimeUnit
+import kotlin.system.measureNanoTime
+
+fun measureMilliAndNanoTime(block: () -> Unit): Double {
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/internal/utilities/TimingUtils.kt:6]: kotlin has a function for this, do we actually benefit from this?

File: core/src/test/kotlin/net/corda/core/internal/profiling/CacheTracingTest.kt [hunk 6]
```diff
@@ -0,0 +1,106 @@
+package net.corda.core.internal.profiling
```
Comment (mnesbit) [inline core/src/test/kotlin/net/corda/core/internal/profiling/CacheTracingTest.kt:1]: This does look superfluous to core and yes should be in an :enterprise-common, or :node-api in ENT till we tackle that.

File: core/src/main/kotlin/net/corda/core/internal/ConcurrentBox.kt (hunk 1 above)
Comment (mnesbit) [reply core/src/main/kotlin/net/corda/core/internal/ConcurrentBox.kt:12]: I second that

File: core/src/main/kotlin/net/corda/core/internal/utilities/TimingUtils.kt (hunk 5 above)
Comment (blsemo) [reply core/src/main/kotlin/net/corda/core/internal/utilities/TimingUtils.kt:6]: I added this function when we did the war room about where we looked into the lost messages - it is used in one place in the bridge to measure time in milliseconds with micro/nano seconds as decimal places and log it at debug level. We might as well remove it.

**Pull request closed without merging**